src/clients/go/tb_client.go

```diff
@@ -203,6 +209,7 @@ func (c *c_client) doRequest(
 
 	// Wait for the request to complete.
 	<-req.ready
+	pinner.Unpin()
```

Comment:
I have faced this question before, and my reasoning was that we didn't need to pin memory because the reference is held during the duration of the gorountine, that is, it is not a callback in the sense of two independent calls, but just a “wait for the channel signal”.

So, why this assumption is not true and the GC is removing the reference? Is there some kind of compiler state machine under the hood, like in async functions?

Reply:
From the language-lawyer perspective, the cgo docs are pretty clear that:

* all things must be pinned
* arguments are pinned implicitly only during a single FFI call

while this  ref is scoped to a go function call, it isn't scoped to a single FFI call, so we are in violation here. And running `cgocheck2` tells pretty much this:

```
write of unpinned Go pointer 0xc000012ab0 to non-Go memory 0x7f7a69b1bfd8
fatal error: unpinned Go pointer stored into non-Go memory
```

as to why this actually blows up at runtime (https://github.com/tigerbeetle/tigerbeetle/pull/1932) I am actually not sure!

One explanation I have:

* the `req` object is allocated on the stack of the goroutine
* when we do `<-`, the goroutine is suspended
* when the goroutine is suspending, the GC moves its stack (I think _stacks_ in Go are movable)
* the pointer now points to "fun"